atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/cache/LockWatchValueScopingCacheImpl.java

```diff
@@ -199,16 +208,18 @@ private synchronized void updateStores(TransactionsLockWatchUpdate updateForTran
         Multimap<Sequence, StartTimestamp> reversedMap = createSequenceTimestampMultimap(updateForTransactions);
 
         // Without this block, updates with no events would not store a snapshot.
-        currentVersion
-                .map(LockWatchVersion::version)
-                .map(Sequence::of)
-                .ifPresent(sequence ->
-                        snapshotStore.storeSnapshot(sequence, reversedMap.get(sequence), valueStore.getSnapshot()));
+        currentVersion.map(LockWatchVersion::version).map(Sequence::of).ifPresent(sequence -> {
+            synchronized (this) {
+                snapshotStore.storeSnapshot(sequence, reversedMap.get(sequence), valueStore.getSnapshot());
+            }
+        });
```

Comment:
I think you need to remove the `synchronized` on this method!

Reply:
ok, pushing the synchronized down to finer grained level

Reply:
I think we can keep the method synchronized as it is private and only called once from a method that already holds the mutex. The synchronized in the lambda is necessary to satisfy the `@GuardedBy` given that the compiler doesn't know the lambda won't escape and run on another thread.

Reply:
ahh, interesting. And since synchronised is reentrant, this should cause minimal / no overhead while making the `@GuardedBy` check more useful. Seems reasonable then.